### PR TITLE
perf: add dLLM consumer-sufficient TP vocab state

### PR DESCRIPTION
**Pull request by Kokoro2336**
## Motivation

Upstream PR: #29970 

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35591242234](https://github.com/sgl-project/sglang/actions/runs/35591242234)
Latest PR Test (Extra): :x: [Run #35591242037](https://github.com/sgl-project/sglang/actions/runs/35591242037)
Latest PR Test (AMD ROCm 10): :x: [Run #35591242246](https://github.com/sgl-project/sglang/actions/runs/35591242246)

### CODE REVIEW

File: python/sglang/srt/dllm/algorithm/base.py [hunk 1]
```diff
@@ -119,7 +120,7 @@ def _run_sync(
         if _is_npu:
             forward_batch.mark_forward_metadata_ready()
         for _ in range(self.max_steps(self.block_size)):
-            done = self.step(forward_batch, out.logits_output.full_logits, states)
+            done = self.step(forward_batch, out.logits_output, states)
```
Comment (rwang5203) [inline python/sglang/srt/dllm/algorithm/base.py]: `Gemma4Renoise.step` still expects a tensor, so the new result-object argument crashes generation even with compact state disabled. Could you adapt that consumer and cover both decoding loops?

File: python/sglang/srt/dllm/tp_local_vocab_state.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        argmax_ids_by_rank,
+        max_id,
+    )
+    best_ids = torch.min(candidate_ids, dim=0).values
+    merged_logsumexp = torch.logsumexp(logsumexp_by_rank, dim=0)
+    return VocabState(
+        max_values=best_values,
+        argmax_ids=best_ids.long(),
+        logsumexp=merged_logsumexp,
+        max_probs=torch.exp(best_values - merged_logsumexp),
```
Comment (rwang5203) [inline python/sglang/srt/dllm/tp_local_vocab_state.py]: Rounded absolute logsumexp can reverse confidence ordering, selecting a lower-confidence position during LowConfidence fallback. Consider merging max-shifted exponential sums and testing near-tied selection against dense softmax.

File: test/srt/dllm/test_tp_local_vocab_state.py [hunk 3]
```diff
@@ -0,0 +1,929 @@
...
+def _dense_argmax_and_max_prob(
+    logits: torch.Tensor,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    logits = logits.float()
+    max_values, argmax_ids = torch.max(logits, dim=-1)
+    max_probs = torch.exp(max_values - torch.logsumexp(logits, dim=-1))
+    return argmax_ids.long(), max_probs
+
+
+def test_low_confidence_tp_state_matches_dense_logits():
```
Comment (rwang5203) [inline test/srt/dllm/test_tp_local_vocab_state.py]: CI only discovers registered tests, so this file leaves compact-state regressions unprotected. Try moving it into the registered tree and adding registration plus a test entry point.

File: python/sglang/srt/dllm/consumer_state_trace.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            "tp_size": int(tp_size),
+            "rows": rows,
+            "vocab_size": int(vocab_size),
+            "local_vocab_size": int(full_logits.shape[-1])
+            if full_logits is not None and full_logits.ndim > 0
+            else 0,
+            "full_vocab_materialized_bytes": tensor_nbytes(full_logits),
+            "full_vocab_reread_bytes": 0,
+            "compact_state_bytes": 0,
+            "tp_gather_bytes": tensor_nbytes(full_logits) if int(tp_size) > 1 else 0,
```
Comment (rwang5203) [inline python/sglang/srt/dllm/consumer_state_trace.py]: Dense tracing counts FP32 output bytes for BF16 or FP16 gathers, overstating the gathered payload. Consider recording the actual gather buffer's shape and dtype.

File: python/sglang/srt/environ.py [hunk 5]
```diff
@@ -331,6 +331,9 @@ class Envs:
     SGLANG_EXTERNAL_MODEL_PACKAGE = EnvStr("")
     SGLANG_EXTERNAL_MM_MODEL_ARCH = EnvStr("")
     SGLANG_EXTERNAL_MM_PROCESSOR_PACKAGE = EnvStr("")
+    # dLLM: exchange consumer-sufficient statistics instead of full TP logits.
+    SGLANG_DLLM_TP_LOCAL_VOCAB = EnvBool(False)
+    SGLANG_DLLM_TP_LOCAL_VOCAB_PACKED_GATHER = EnvBool(True)
```
Comment (rwang5203) [inline python/sglang/srt/environ.py]: This default-off feature has no shipped usage guidance, leaving users to discover its environment flag and fallback conditions. Could you document the supported launch path and limitations?
